packages/sdk-node/src/authenticateRequest.ts

```diff
@@ -56,3 +81,26 @@ export const decorateResponseWithObservabilityHeaders = (res: ServerResponse, re
   requestState.reason && res.setHeader(constants.Headers.AuthReason, encodeURIComponent(requestState.reason));
   requestState.status && res.setHeader(constants.Headers.AuthStatus, encodeURIComponent(requestState.status));
 };
+
+const isDevelopmentFromApiKey = (apiKey: string): boolean =>
+  apiKey.startsWith('test_') || apiKey.startsWith('sk_test_');
+
+const getRequestUrl = (req: IncomingMessage): URL => {
+  return new URL(req.url as string, `${getRequestProto(req)}://${req.headers.host}`);
+};
+
+const getRequestProto = (req: IncomingMessage): string => {
+  const proto = req.headers[constants.Headers.ForwardedProto] as string;
+  if (!proto) {
+    return req.headers.host && req.headers.host.match('localhost') ? 'http' : 'https';
+  }
+  const index = proto.indexOf(',');
+  return index > -1 ? proto.substring(0, index).trim() : proto.trim();
+};
```

Comment:
I had a lot of trouble getting the request protocol from the [IncomingMessage](https://nodejs.org/api/http.html#class-httpincomingmessage) request.

I ended up looking for the "x-forwarded-proto" header and default to a hacky way to determine the protocol. All requests are considered https unless coming from localhost. 

Any suggestion for a better way to deal with this?

Reply:
What if someone simply added a entry to their `/etc/hosts`  and they host now is `custom.local`, wouldn't that basically break this assumption ?

Reply:
I yesterday gave it a try to find a better way to handle the fallback. I could not find something. Most servers correctly set the X-forwarded-proto header, so we should be good.

Reply:
Discussed offline with @dimkl and @SokratisVidros and we agreed on the following.

- We won't (and can't) change the `req` object's type as some frameworks extend `IncomingMessage` for their custom request objects. A good example is express and its own [Request](https://github.com/expressjs/express/blob/353348a83e6874db7ded27102e91e7b6cad7e89c/lib/request.js#L31).
- Since we're stuck with `IncomingMessage` and there's no easy way to determine the protocol, we'll use the following algorithm:
  1. Check the "X-Forwarded-Proto" header. This should take precedence.
  2. Check if `req.connection.encrypted` is set. Perhaps we're lucky and the `req` object comes in as a `ClientRequest`.
  3. If you can't find the request protocol, throw an error. Fail fast with a descriptive message.
  
Updated the code with the above changes.

Reply:
👍